system-contracts/test/NonceHolder.spec.ts

```diff
@@ -0,0 +1,185 @@
+      const result = await nonceHolder.getDeploymentNonce(nonceHolderAccount.address);
+      expect(result).to.equal(1);
+    });
+  });
+
+  describe("setValueUnderNonce", async () => {
+    it("should revert Nonce value cannot be set to 0", async () => {
+      const accountInfo = [1, 0];
+      const encodedAccountInfo = ethers.utils.defaultAbiCoder.encode(["tuple(uint8, uint8)"], [accountInfo]);
+      await setResult("ContractDeployer", "getAccountInfo", [nonceHolderAccount.address], {
+        failure: false,
+        returnData: encodedAccountInfo,
+      });
+      await expect(nonceHolder.connect(nonceHolderAccount).setValueUnderNonce(124, 0)).to.be.rejectedWith(
+        "Nonce value cannot be set to 0"
+      );
+    });
+
+    it("should revert Previous nonce has not been used", async () => {
+      const accountInfo = [1, 0];
+      const encodedAccountInfo = ethers.utils.defaultAbiCoder.encode(["tuple(uint8, uint8)"], [accountInfo]);
+      await setResult("ContractDeployer", "getAccountInfo", [nonceHolderAccount.address], {
+        failure: false,
+        returnData: encodedAccountInfo,
+      });
+      await expect(nonceHolder.connect(nonceHolderAccount).setValueUnderNonce(123, 111)).to.be.rejectedWith(
+        "Previous nonce has not been used"
+      );
+    });
+
+    it("should emit ValueSetUnderNonce event", async () => {
+      const expectedNonce = await nonceHolder.getMinNonce(nonceHolderAccount.address);
+      await nonceHolder.connect(nonceHolderAccount).incrementMinNonceIfEquals(expectedNonce);
+      const accountInfo = [1, 0];
+      const encodedAccountInfo = ethers.utils.defaultAbiCoder.encode(["tuple(uint8, uint8)"], [accountInfo]);
+      await setResult("ContractDeployer", "getAccountInfo", [nonceHolderAccount.address], {
+        failure: false,
+        returnData: encodedAccountInfo,
+      });
+      await expect(nonceHolder.connect(nonceHolderAccount).setValueUnderNonce(expectedNonce.add(1), 333))
+        .to.emit(nonceHolder, "ValueSetUnderNonce")
+        .withArgs(nonceHolderAccount.address, expectedNonce.add(1), 333);
+      const key = await nonceHolder.getMinNonce(nonceHolderAccount.address);
+      const storedValue = await nonceHolder.connect(nonceHolderAccount).getValueUnderNonce(key);
+      expect(storedValue).to.equal(333);
+    });
+  });
+
+  describe("getValueUnderNonce", () => {
+    it("should get value under nonce", async () => {
+      const key = await nonceHolder.getMinNonce(nonceHolderAccount.address);
+      const storedValue = await nonceHolder.connect(nonceHolderAccount).getValueUnderNonce(key);
+      expect(storedValue).to.equal(333);
+    });
+  });
+
+  describe("isNonceUsed", () => {
+    it("used nonce", async () => {
+      const isUsed = await nonceHolder.isNonceUsed(nonceHolderAccount.address, 1);
+      expect(isUsed).to.equal(true);
+    });
+
+    it("not used nonce", async () => {
```

Comment:
Can you test both cases, when used by `_nonce < getMinNonce(_address)` and when by `nonceValues[addressAsKey][_nonce] > 0`?

Reply:
done